src/rust/receiver-mock/src/router.rs

```diff
@@ -85,6 +112,103 @@ pub async fn handler_metrics_ips(app_state: web::Data<AppState>) -> impl Respond
     HttpResponse::Ok().body(out)
 }
 
+// Return consumed metrics. Endpoint handled by this handler will accept URL query
+// which is a key value list of label names and values that the metric should contain
+// in order be returned.
+// Label values can be ommitted in which case only presence of a particular label
+// will be checked, not its value in the filtered sample.
+// `__name__` is handled specially as it will be matched against the metric name.
+//
+// Exemplar usage of this endpoint:
+//
+// $ curl -s localhost:3000/metrics-samples\?__name__=apiserver_request_total\&cluster | jq .
+// [
+//     {
+//       "metric": "apiserver_request_total",
+//       "value": 124,
+//       "labels": {
+//         "prometheus_replica": "prometheus-release-test-1638873119-ku-prometheus-0",
+//         "_origin": "kubernetes",
+//         "component": "apiserver",
+//         "service": "kubernetes",
+//         "resource": "events",
+//         "code": "422",
+//         "instance": "172.18.0.2:6443",
+//         "group": "events.k8s.io",
+//         "namespace": "default",
+//         "verb": "POST",
+//         "scope": "resource",
+//         "endpoint": "https",
+//         "version": "v1",
+//         "job": "apiserver",
+//         "cluster": "microk8s",
+//         "prometheus": "ns-test-1638873119/release-test-1638873119-ku-prometheus"
+//       },
+//       "timestamp": 163123123
+//     }
+//   ]
+//
+pub async fn handler_metrics_samples(
+    app_state: web::Data<AppState>,
+    web::Query(params): web::Query<HashMap<String, String>>,
+    opts: web::Data<options::Options>,
+) -> impl Responder {
+    if !opts.store_metrics {
+        return HttpResponse::NotImplemented().body("");
+    }
+
+    let x: HashSet<Sample> = app_state
+        .metrics_samples
+        .lock()
+        .unwrap()
+        .clone()
```

Comment:
Yup, it seems to:

```
cannot move out of dereference of `std::sync::MutexGuard<'_, std::collections::HashSet<metrics::Sample>>`

move occurs because value has type `std::collections::HashSet<metrics::Sample>`, which does not implement the `Copy` traitrustc(E0507)
```

Reply:
Ah right, the set owns the elements, and `collect` moves ownership to the new set.

In that case, I'd suggest moving the cloning to after the filter:
```rust
    let x: HashSet<Sample> = app_state
        .metrics_samples
        .lock()
        .unwrap()
        .iter()
        .filter(|sample| {
            true
        })
        .cloned()
        .collect();
```
That way you only clone the elements you put in the output set, rather than all of them.

Reply:
Thanks! Changed